### PR TITLE
cmake install improvments

### PR DESCRIPTION
**Pull request by orbea**
This PR has two related install fixes that can help distros.

* GNU install standards state that `datadir` should be used for program specific data and `datarootdir` should be used for other data files such as icons or desktop files. However nxengine-evo installs everything to `CMAKE_INSTALL_DATADIR`. I changed the non-game data to use `CMAKE_INSTALL_DATAROOTDIR` instead. By default `CMAKE_INSTALL_DATADIR` sources the value of `CMAKE_INSTALL_DATAROOTDIR`, but its important to be able to install them into separate locations.
* For pc builds this adds a `DATADIR` define to indicate the locations of the data files, it uses the value of `${CMAKE_INSTALL_FULL_DATADIR}/nxengine/data/`. This allows the user to install nxengine-evo to arbitrary locations. For example `/usr/share/games/nxengine` or even `$HOME/games/nxengine`.

### CODE REVIEW

File: src/ResourceManager.cpp [hunk 1]
```diff
@@ -97,6 +97,11 @@ std::string ResourceManager::getLocalizedPath(const std::string &filename)
     _paths.push_back("../share/nxengine/data/mods/" + _mod + "/lang/" + std::string(settings->language) + "/" + filename);
     _paths.push_back("../share/nxengine/data/mods/" + _mod + "/" + filename);
   }
+  #if defined(DATADIR)
```
Comment (isage) [inline src/ResourceManager.cpp:100]: That should be replicated for mods and for getPathForDir

**Review by isage — CHANGES_REQUESTED** [review]

File: src/ResourceManager.cpp (hunk 1 above)
Comment (orbea) [reply src/ResourceManager.cpp:106]: Done, thanks for pointing this out.

**Review by isage — APPROVED** [review]

**Comment by isage** [comment]
@ntninja @timsueberkrueb can you take a look and check that it won't break packaging?

File: src/ResourceManager.cpp [hunk 2]
```diff
@@ -88,6 +88,11 @@ std::string ResourceManager::getLocalizedPath(const std::string &filename)
 
   if (!_mod.empty())
   {
+    #if defined(DATADIR)
+      _paths.push_back(std::string(DATADIR) + "mods/" + _mod + "/lang/" + std::string(settings->language) + "/" + filename);
+      _paths.push_back(std::string(DATADIR) + "mods/" + _mod + "/" + filename);
+    #endif
+
```
Comment (ntninja) [inline src/ResourceManager.cpp:95]: I believe this should replace the hardcoded `"/usr/share"` below, rather than make the list even longer. Also if there is a way to query the equivalent of `"/usr/local/share"` from CMake that should replace *those* lines as well.

That way nxevo won't try to access unrelated paths should the distributor decide on a different pathing scheme.

File: src/ResourceManager.cpp [hunk 3]
```diff
@@ -202,11 +213,19 @@ std::string ResourceManager::getPathForDir(const std::string &dir)
 
   if (!_mod.empty())
   {
+    #if defined(DATADIR)
+      _paths.push_back(std::string(DATADIR) + "mods/" + _mod + "/" + dir);
+    #endif
```
Comment (ntninja) [inline src/ResourceManager.cpp:218]: … same here of course.

**Review by ntninja — CHANGES_REQUESTED** [review]
I believe this should supersede the previous implementation, not make it more unwieldy. I haven't tested these changes yet, but I believe they are fine for Flatpak/AppImage.

File: src/ResourceManager.cpp (hunk 2 above)
Comment (isage) [reply src/ResourceManager.cpp:95]: `/usr/local/share` probably won't be needed, as that will replace it too

Comment (orbea) [reply src/ResourceManager.cpp:95]: Should there be a fallback in case `DATADIR` is not defined? I'm not sure that would ever happen.

Comment (orbea) [reply src/ResourceManager.cpp:108]: After thinking about it more I think the correct behavior would be to always check `/usr/local/share/nxengine/data/` first unless `DATADIR` is already set to that value.

I am not sure if there should be a fallback if `DATADIR` is not defined, but to be safe I did that for now. I can change this if other's disagree.

I rebased and updated the PR accordingly.

Comment (isage) [reply src/ResourceManager.cpp:108]: `/usr/local/share` is a default for `${CMAKE_INSTALL_FULL_DATADIR}`
User can provide custom data in `$HOME/.local/...`

Comment (orbea) [reply src/ResourceManager.cpp:108]: Good point, but its a bit more than just a cmake default. Its part of the FHS standard. Using `$HOME` would only affect that one user and using `/usr/local` would affect all users on the system.

> The /usr/local hierarchy is for use by the system administrator when installing software locally. It needs to be safe from being overwritten when the system software is updated. It may be used for programs and data that are shareable amongst a group of hosts, but not found in /usr.
>
> Locally installed software must be placed within /usr/local rather than /usr unless it is being installed to replace or upgrade software in /usr.

https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch04s09.html

**Comment by timsueberkrueb** [comment]
This breaks the snap, but I can update it accordingly.

File: src/ResourceManager.cpp (hunk 2 above)
Comment (isage) [reply src/ResourceManager.cpp:108]: Well, "not found in usr". That's the point, it's either `/usr` or `/usr/local`.

Comment (orbea) [reply src/ResourceManager.cpp:108]: > Well, "not found in usr". That's the point, it's either `/usr` or `/usr/local`.

Imagine a scenario where the user has a nxengine-evo package that installs files into `/usr/share/nxengine/`, but they wish to provide new data files to replace the packages files as explained in the second block of text. Their choices would be to install them into `$HOME/.local` for every applicable user of if they want it to affect all users they would use `/usr/local`.

I suppose they could build and install nxengine-evo entirely into `/usr/local`, but perhaps if all they care about is testing modifications to the game data it might be easier to only replace that.

> Locally installed software must be placed within /usr/local rather than /usr unless it is being installed to replace or upgrade software in /usr.

Comment (isage) [reply src/ResourceManager.cpp:108]: I still think that its a very hypothetical scenario. We can always add `/usr/local` back if there's high demand for that

Comment (orbea) [reply src/ResourceManager.cpp:101]: I will defer to your judgement and made the changes.

**Comment by isage** [comment]
Merged. @timsueberkrueb your turn

**Comment by timsueberkrueb** [comment]
What is the recommended way for me to override `DATADIR` given that the build-time location is different from the runtime location? 
In its build process, `snapcraft` locally installs using `DESTDIR`. Files under `$DESTDIR/usr` will end up at runtime at `/snap/nxengine-evo/current/usr`.

File: CMakeLists.txt [hunk 4]
```diff
@@ -103,8 +102,9 @@ add_definitions("-Wall")
 add_executable(nx ${SOURCES})
 
 IF(PLATFORM STREQUAL "pc")
-set_property(TARGET nx PROPERTY OUTPUT_NAME nxengine-evo)
+    set_property(TARGET nx PROPERTY OUTPUT_NAME nxengine-evo)
     add_definitions("-std=c++11")
+    add_definitions(-DDATADIR="${CMAKE_INSTALL_FULL_DATADIR}/nxengine/data/")
```
Comment (ntninja) [inline CMakeLists.txt:107]: Looking at the documentation it appears that it would be better to use `add_compile_definitions(DATADIR="…")` here which doesn't have weird flags mangling behaviour, requires CMake 3.12 though.

File: CMakeLists.txt [hunk 5]
```diff
@@ -103,8 +102,9 @@ add_definitions("-Wall")
 add_executable(nx ${SOURCES})
 
 IF(PLATFORM STREQUAL "pc")
-set_property(TARGET nx PROPERTY OUTPUT_NAME nxengine-evo)
+    set_property(TARGET nx PROPERTY OUTPUT_NAME nxengine-evo)
     add_definitions("-std=c++11")
```
Comment (ntninja) [inline CMakeLists.txt:106]: This should use the more semantic `set_property(TARGET nx PROPERTY CXX_STANDARD 11)` instead.

**Comment by ntninja** [comment]
CMake's [GNUInstallDirs](https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html) doesn't support doing that.

That aside: I thought that  snaps are required to be location-independent, so hard-coding `/snap/nxengine-evo/current/usr/share/nxengine` is wrong just like the previous `/usr/share/nxengine` was, right? So I guess, what you really need is the previous `../share/nxengine` that became conditional for some reason?

Q to @isage: Is the hard-coded `/usr/share/nxengine` & `/usr/local/share/nxengine` actually needed on any of the non-PC targets?

Assuming the answers to the above questions are yes, yes, no respectively, I'd suggest doing this:

  1. Drop the hard-coded `/usr/share/nxengine` & `/usr/local/share/nxengine` paths from the source entirely
  2. Add a new `PORTABLE_INSTALL` boolean variable to CMake that prevents the `-DDATADIR` from being defined if true
  3. Use the above in the Snap build

**Comment by isage** [comment]
Nope, /usr/(local/)/share is only for linux/bsd. Osx doesn't need that and
haiku, afaik, too

**Comment by ntninja** [comment]
@isage: And do console-ports need it? I just checked Haiku in a VM and it appears as if it will work with `GNUInstallDirs` if `-DCMAKE_INSTALL_DIR` is set to `/boot/system/apps/AppName` – on the other hand `/usr` doesn't even exist there so the previous code was wrong. Probably best to just use the relative path approach on this platform as well.

TL;DR: If consoles don't need it these paths than the can be dropped.

**Comment by isage** [comment]
Consoles don't need it, yeah

**Comment by orbea** [comment]
> What is the recommended way for me to override `DATADIR` given that the build-time location is different from the runtime location?
> In its build process, `snapcraft` locally installs using `DESTDIR`. Files under `$DESTDIR/usr` will end up at runtime at `/snap/nxengine-evo/current/usr`.

Why not set `CMAKE_INSTALL_PREFIX` to `/snap/nxengine-evo/current/usr` and then install using `DESTDIR=$DESTDIR`? I don't know much about snap so I might be missing something.

```
mkdir build
cd build
cmake -DCMAKE_INSTALL_PREFIX=/snap/nxengine-evo/current/usr ..
make
make install DETDIR=$DESTDIR
```

**Comment by orbea** [comment]
> Consoles don't need it, yeah

None of this code should affect consoles.

https://github.com/nxengine/nxengine-evo/blob/389e371700e67b77f8eb87cd2d0bc91578730b12/CMakeLists.txt#L104
https://github.com/nxengine/nxengine-evo/blob/389e371700e67b77f8eb87cd2d0bc91578730b12/src/ResourceManager.cpp#L75

File: CMakeLists.txt (hunk 4 above)
Comment (orbea) [reply CMakeLists.txt:107]: Its currently set at `2.6.

https://github.com/orbea/nxengine-evo/blob/b12351b39020f3f134ea51daafd9ef8c199193aa/CMakeLists.txt#L1

Slackware 14.2 is stuck at `3.5.2`, but then there hasn't been a stable release since 2016...

I wrote it like that because that is how other parts of the cmake scripts were written.

File: CMakeLists.txt (hunk 5 above)
Comment (orbea) [reply CMakeLists.txt:106]: I just fixed the indent in that line because it made it easier for me to read the relevant code.

**Comment by timsueberkrueb** [comment]
> So I guess, what you really need is the previous ../share/nxengine that became conditional for some reason

That is indeed what the snap previously relied on. 

Regarding location-independence I'm fairly sure that `/snap/<appname>/current/` can be relied upon (iirc, this is in fact why classic - i.e. non-confined - snaps are not supported on installations where `/snap` does not point to the snaps mount point directory). Unfortunately, I could not find a formal statement regarding that in the docs, so I may be wrong.

> Why not set CMAKE_INSTALL_PREFIX to /snap/nxengine-evo/current/usr and then install using DESTDIR=$DESTDIR? I don't know much about snap so I might be missing something.

While I can certainly do something like this, afaik this will require me to work against snapcraft which expects the files to be installed to `$DESTDIR` rather than `$DESTDIR/snap/nxengine-evo/current/usr`, i.e. without doing some trickery, the files would end up at runtime in `/snap/nxengine-evo/current/snap/nxengine-evo/current/usr`. I might be wrong, though.

**Comment by orbea** [comment]
Do you have a link to the current nxengine-evo snap build repo?

Using `DESTDIR` is for staged installs, its a temporary directory that is then used to create packages that are then installed to the filesystem using the value of `CMAKE_INSTALL_<dir>` with `DESTDIR` removed.

I suspect snap would have problems in a lot of programs if they rely on programs being installed and used from the `DESTDIR`,  remember that its a variable known to the `ninja` or `make` implementation and not necessarily `cmake` itself.

**Comment by timsueberkrueb** [comment]
Yeah, `snapcraft` uses [these stages](https://snapcraft.io/docs/parts-lifecycle) and [here is what the CMake plugin does](https://github.com/snapcore/snapcraft/blob/master/snapcraft/plugins/v2/cmake.py#L98).

File: CMakeLists.txt (hunk 5 above)
Comment (ntninja) [reply CMakeLists.txt:106]: Oh sorry, this was a general comment, I don't blame you for this or anything! :slightly_smiling_face:

**Comment by orbea** [comment]
@timsueberkrueb Thanks, but I'm confused why this breaks snap? Can't you just use `-DDATADIR=/snap/nxengine-evo/current/usr/share`? Does snap even  use `CMAKE_INSTALL_PREFIX`? It should already work if snap correctly sets `CMAKE_INSTALL_PREFIX` to the actual installed path. Or is this entirely arbitrary at runtime?

If needed I think you would want to use either `cmake-parameters` or `configflags`.

https://snapcraft.io/docs/cmake-plugin

**Comment by timsueberkrueb** [comment]
Setting `-DDATADIR=/snap/nxengine-evo/current/usr/share` doesn't seem to work:

```
   Manually-specified variables were not used by the project:

    DATADIR
```

Setting `CMAKE_INSTALL_PREFIX` to `/snap/nxengine-evo/current/usr/share` will lead to the following build-time directory structure:

```
parts/nxengine-evo/install/snap
└── nxengine-evo
    └── current
        └── usr
            └── share
                ├── bin
                └── share
```

Snapcraft expects `parts/nxengine-evo/install` to be the install root. Hence, this results in the runtime path `/snap/nxengine-evo/current/snap/nxengine-evo/current/usr`. That's what I meant by "working against snapcraft". Of course, I can copy things around manually if need be. 

Though I think it would be preferable to just be able to set the runtime `DATADIR` from the command line.

Btw, I've [posted in the snapcraft forum](https://forum.snapcraft.io/t/guarantees-regarding-snap-appname-current/22397), regarding @ntninja's question whether building the string into the binary is fine after all.

**Comment by timsueberkrueb** [comment]
It was suggested in the forum that we could also use [snap's layout feature ](https://snapcraft.io/docs/snap-layouts).

**Comment by orbea** [comment]
> Setting `-DDATADIR=/snap/nxengine-evo/current/usr/share` doesn't seem to work:

Sorry, I meant to type `-DCMAKE_INSTALL_DATADIR=/snap/nxengine-evo/current/usr/share` .

**Comment by timsueberkrueb** [comment]
Given #212 is needed for AppImage anyway, as @ntninja suggested above, I think it makes most sense to just also use that for the snap, then :)
